impeller/tessellator/tessellator.cc

```diff
@@ -224,4 +224,349 @@ void DestroyTessellator(TESStesselator* tessellator) {
+  // by finding the midpoint of the line of the first slice and compare
+  // its distance from the center of the circle to the radius. We will aim
+  // to have the length of that bisector to be within |kCircleTolerance|
+  // from the radius in pixels.
+  //
+  // Each vertex will appear at an angle of:
+  //   theta(i) = (kPi / 2) * (i / N)  // for i in [0..N]
+  // with each point falling at:
+  //   point(i) = r * (cos(theta), sin(theta))
+  // If we consider the unit circle to simplify the calculations below then
+  // we need to scale the tolerance from its absolute quantity into a unit
+  // circle fraction:
+  //   k = tolerance / radius
+  // Using this scaled tolerance below to avoid multiplying by the radius
+  // throughout all of the math, we have:
+  //   first point = (1, 0)   // theta(0) == 0
+  //   theta = kPi / 2 / N    // theta(1)
+  //   second point = (cos(theta), sin(theta)) = (c, s)
+  //   midpoint = (first + second) * 0.5 = ((1 + c)/2, s/2)
+  //   |midpoint| = sqrt((1 + c)*(1 + c)/4 + s*s/4)
+  //     = sqrt((1 + c + c + c*c + s*s) / 4)
+  //     = sqrt((1 + 2c + 1) / 4)
+  //     = sqrt((2 + 2c) / 4)
+  //     = sqrt((1 + c) / 2)
+  //     = cos(theta / 2)     // using half-angle cosine formula
+  //   error = 1 - |midpoint| = 1 - cos(theta / 2)
+  //   cos(theta/2) = 1 - error
+  //   theta/2 = acos(1 - error)
+  //   kPi / 2 / N / 2 = acos(1 - error)
+  //   kPi / 4 / acos(1 - error) = N
+  // Since we need error <= k, we want divisions >= N, so we use:
+  //   N = ceil(kPi / 4 / acos(1 - k))
+  //
+  // Math is confirmed in https://math.stackexchange.com/a/4132095
+  // (keeping in mind that we are computing quarter circle divisions here)
+  // which also points out a performance optimization that is accurate
+  // to within an over-estimation of 1 division would be:
+  //   N = ceil(kPi / 4 / sqrt(2 * k))
+  // Since we have precomputed the divisions for radii up to 1024, we can
+  // afford to be more accurate using the acos formula here for larger radii.
+  double k = Tessellator::kCircleTolerance / pixel_radius;
+  return ceil(kPiOver4 / std::acos(1 - k));
+}
+
+void Tessellator::Trigs::init(size_t divisions) {
+  if (!trigs_.empty()) {
+    return;
+  }
+
+  // Either not cached yet, or we are using the temp storage...
+  trigs_.reserve(divisions + 1);
+
+  double angle_scale = kPiOver2 / divisions;
+
+  trigs_.emplace_back(1.0, 0.0);
+  for (size_t i = 1; i < divisions; i++) {
+    trigs_.emplace_back(Radians(i * angle_scale));
+  }
+  trigs_.emplace_back(0.0, 1.0);
+}
+
+Tessellator::Trigs Tessellator::GetTrigsForDivisions(size_t divisions) {
+  return divisions < Tessellator::kCachedTrigCount
+             ? Trigs(precomputed_trigs_[divisions], divisions)
+             : Trigs(divisions);
+}
+
+using TessellatedVertexProc = Tessellator::TessellatedVertexProc;
+using VertexGenerator = Tessellator::VertexGenerator;
+
+std::unique_ptr<VertexGenerator> Tessellator::FilledCircle(
```

Comment:
I couldn't return the sub-classes until I did something like this. Maybe there is magic to do that, but I couldn't find it and then found an answer on the web saying that you can't return a sub-class for a super-class and need to use a managed pointer to do it.

Reply:
This would have to return FilledCircleGenerator instead of VertexGenerator

Reply:
Which would involve exposing the utility classes. Also, some of them might want to return one of a set of various sub-classes depending on the data.

Reply:
But now we're doing a new heap allocation per piece of data we're processing, which mid-long term we'll need to unwind to deal with scudo (android malloc implementation) being a slow allocator in general.

Reply:
Look at StrokedCircle which might return a FilledCircle implementation if the conditions are right. Also RoundCap and FilledEllipse have variant returns.

I could probably do this by forcing the caller to make these decisions and these methods always return a specific value even if you've chosen poorly, but in the future the decision might be more complex than "that value is 0". I would also have to make all of the implementation classes public even though they can only be instantiated from inside the Tessellator (and contain data that isn't really public-facing).

Any other ideas? I see your point and I want to achieve that (it was a feature of the previous packaging of the mechanism). Unfortunately the prior mechanism required you to pass in data during instantiation and then pass in mostly the same data again for the case-specific generator method.

Reply:
One thing I'm struggling with is multiple players that each want to manage the process. So, making a Buffer really wants to know up-front how many points there so the process ends up being "I want to do X, how many points will that generate?" followed by "Me again, I want to do X, can you give me the points now?". I created the Generator to encapsulate both answers inside a single object.

Hmmm, maybe the Generator could have union of all of the data and pluggable pointer to a static function, so each case ends up being "VG(trigs, func, data)" and they are stored in a stack allocated struct/class?

Reply:
We could have it contain a union of all possible data that any of them might need and then just sub in a generator proc. That could be limiting in the long run, but for now we only have half a dozen primitives and they all have several pieces of data in common. Then the various methods just have to figure out which generator proc to supply, and that proc would only use some of the data. (Or an actual `union{}`, but that might leave us open to splicing the data rather than just filling it all out even for the simple operations that only need a couple of the fields?).

For example, everything is a round rect, it has 4 center points and 2 radii, a stroke width and possibly an along and an across vector . But, if all four center points are the same (because it was instantiated from `Tessellator::FilledCircle`) then the proc installed would just grab one of them and ignore the other 3. The along and across vectors might be `{1,0}` and `{0,1}` for most use cases except for a `RoundCapLine` so they are ignored by all procs except for the one specifically designed to produce round caps (and even then it could compute those from the first 2 points). Even arc tessellation only adds 2 angle values. So, we'd have:

```
  class VertexGenerator {
   public:
    const PrimitiveType triangle_type;
    const size_t vertex_count;

    void GenerateVertices(const PointProc& proc) { impl_(*this, proc); }
    // or a switch statement and a shape_type enum?

   private:
    const function<void(const VertexGenerator& generator, const PointProc& proc)>& impl_;

    const Point corners_[4];
    const Size radii_;
    const Scalar stroke_width_;
    // const Radians start_angle_;
    // const Radians end_angle_;  (or extent)
  }
```

Not every use case needs all of those values, but the stack allocated object will have default values for all of them and assigned values for the ones needed for a given use case. That's 4 Points a Size and a (or 1-3) Scalar(s), but it saves heap allocation. And I don't think it needs a destructor.

Then:
- FilledCircle sets all corners to the center, and the radii to `{radius, radius}`
- StrokedCircle sets all of those and the stroke_width;
- RoundCapLine sets the corners to p0, p1, + dups and the radii to `{half_line_width, half_line_width}`
  - Potentially it could also precompute the along and across vectors, but these aren't used more than once in a typical use case so that isn't really necessary.
- FilledEllipse sets all corners to the center of the oval bounds and the radii_ to half the size of the bounds.
- FilledRoundRect sets the corners to the interior corners of the round rect and the radii_ to the corner radii (only works for same radii at all 4 corners)
- FilledArc would only need the start and end angles over FilledCircle.

Reply:
I don't think you necessarily need the VertexGenerator to have the function pointer, it could have an enum or state bit and a shared_ptr to the original tessellator instance. Then when  you call generate vertices, you jump back into the shared_ptr tessellator with the state you need.

Reply:
I'm using:

```
    typedef void GeneratorProc(const Trigs& trigs,
                               const Data& data,
                               const TessellatedVertexProc& proc);

  // ...

  static void GenerateFilledCircle(const Trigs& trigs,
                                   const VertexGenerator::Data& data,
                                   const TessellatedVertexProc& proc);

  static void GenerateStrokedCircle(const Trigs& trigs,
                                    const VertexGenerator::Data& data,
                                    const TessellatedVertexProc& proc);

  static void GenerateRoundCapLine(const Trigs& trigs,
                                   const VertexGenerator::Data& data,
                                   const TessellatedVertexProc& proc);

  static void GenerateFilledEllipse(const Trigs& trigs,
                                    const VertexGenerator::Data& data,
                                    const TessellatedVertexProc& proc);
```

Those static Generator* methods do the actual tessellation work and get set up using something like:

```
    return VertexGenerator(Tessellator::GenerateRoundCapLine,
                           GetTrigsForDivisions(divisions),
                           PrimitiveType::kTriangleStrip, 4,
                           VertexGenerator::Data{
                               .reference_centers = {p0, p1},
                               .radii = {radius, radius},
                               .half_width = -1.0f,
                           });
```